### PR TITLE
Ensure Safe Pointer Usage in `ProgItem` class

### PR DESCRIPTION
**Pull request by KOUGBLENOU**
This PR completes https://github.com/statinf-otawa/elm/pull/1 with the goal of fixing memory leaks described in https://github.com/statinf-software/rocqstat-ng/issues/1354.

### CODE REVIEW

File: src/prog/prog_ProgItem.cpp [hunk 1]
```diff
@@ -43,7 +43,7 @@ ProgItem::~ProgItem(void) {
  */
 ProgItem *ProgItem::next(void) const {
 	ProgItem *result = (ProgItem *)inhstruct::DLNode::next(); 
-	if(!result || result->atEnd())
+	if(result && result->atEnd())
```
Comment (jordr) [inline src/prog/prog_ProgItem.cpp]: This returns 0 when the next node exists, which is the opposite of the intended behaviour, and probably ends every analysis early. I suppose this is how you end up with no memory leaks...

File: src/prog/prog_ProgItem.cpp [hunk 2]
```diff
@@ -56,7 +56,7 @@ ProgItem *ProgItem::next(void) const {
  */
 ProgItem *ProgItem::previous(void) const {
 	ProgItem *result = (ProgItem *)inhstruct::DLNode::previous(); 
-	if(!result || result->atBegin())
+	if(result && result->atBegin())
```
Comment (jordr) [inline src/prog/prog_ProgItem.cpp]: same

**Review by jordr — CHANGES_REQUESTED** [review]

File: src/prog/prog_ProgItem.cpp (hunk 1 above)
Comment (KOUGBLENOU) [reply src/prog/prog_ProgItem.cpp]: Good catch.

>  I suppose this is how you end up with no memory leaks...
No, with or without this change, we end up with no memory leak.

As described above, in this PR, I add an extra check on result pointer before calling `atEnd()` or `atBegin()` method.

File: src/prog/prog_ProgItem.cpp [hunk 3]
```diff
@@ -43,7 +43,7 @@ ProgItem::~ProgItem(void) {
  */
 ProgItem *ProgItem::next(void) const {
 	ProgItem *result = (ProgItem *)inhstruct::DLNode::next(); 
-	if(!result || result->atEnd())
+	if(!result || (result && result->atEnd()))
```
Comment (jordr) [inline src/prog/prog_ProgItem.cpp:46]: `(!A || (A && B))` simplifies to `(!A || B)`, so what you added changes nothing. C++ uses lazy evaluation, so if `!result`, then `result->atEnd()` will not be evaluated.

**Review by jordr — CHANGES_REQUESTED** [review]

File: src/prog/prog_ProgItem.cpp (hunk 3 above)
Comment (KOUGBLENOU) [reply src/prog/prog_ProgItem.cpp:46]: Ooooh I see. Let's close this PR then and focus on https://github.com/statinf-otawa/elm/pull/1.

**Pull request closed without merging**